drivers/i2c/i2c_esp32.c

```diff
@@ -0,0 +1,781 @@
+	status = *config->regs.intr_raw;
+	if (status & (I2C_ARBITRATION_LOST_INT_RAW | I2C_ACK_ERR_INT_RAW)) {
+		return -EIO;
+	}
+	if (status & I2C_TIME_OUT_INT_RAW) {
+		return -ETIMEDOUT;
+	}
+
+	if (wait_cmd) {
+		while (!wait_cmd->done) {
+			r = i2c_esp32_spin_yield(&counter);
+			if (r < 0) {
+				return r;
+			}
+		}
+	}
+
+	/* Wait for I2C bus to finish its business */
+	while (*config->regs.status & I2C_BUS_BUSY) {
+		r = i2c_esp32_spin_yield(&counter);
+		if (r < 0) {
+			return r;
+		}
+	}
+
+	return 0;
+}
+
+static int i2c_esp32_read_msg(struct device *dev, u16_t addr,
+			      struct i2c_msg msg)
+{
+	const struct i2c_esp32_config *config = dev->config->config_info;
+	volatile struct i2c_esp32_cmd *cmd_base = (struct i2c_esp32_cmd *)config->regs.cmd_base;
+	volatile struct i2c_esp32_cmd *cmd_after_write;
+	volatile struct i2c_esp32_cmd *cmd = cmd_base;
+	struct i2c_esp32_data *data = dev->driver_data;
+	u32_t i;
+	int r;
+
+	/* Set the R/W bit to R */
+	addr |= BIT(0);
+
+	i2c_esp32_reset_fifo(config);
+
+	/* Start transaction */
+	if (msg.flags & I2C_MSG_RESTART) {
+		*cmd++ = (struct i2c_esp32_cmd) {
+			.opcode = I2C_ESP32_OP_RSTART
+		};
+	}
+
+	/* Send address */
+	*config->regs.fifo = addr & I2C_FIFO_RDATA;
+	if (data->config.bits.use_10_bit_addr) {
+		*config->regs.fifo = (addr >> 8) & I2C_FIFO_RDATA;
+	}
+	*cmd++ = (struct i2c_esp32_cmd) {
+		.opcode = I2C_ESP32_OP_WRITE,
+		.ack_en = true,
+		.num_bytes = 1 + data->config.bits.use_10_bit_addr,
+	};
+
+	for (cmd_after_write = cmd; msg.len; cmd = cmd_base) {
+		volatile struct i2c_esp32_cmd *wait_cmd = NULL;
+		u32_t to_read = max(I2C_ESP32_BUFFER_SIZE, msg.len - 1);
+
+		/* Might be the last byte, in which case, `to_read` will
+		 * be 0 here.  See comment below.
+		 */
+		if (to_read) {
+			*cmd++ = (struct i2c_esp32_cmd) {
+				.opcode = I2C_ESP32_OP_READ,
+				.num_bytes = to_read,
+			};
+		}
+
+		/* I2C master won't acknowledge the last byte read from the
+		 * slave device.  Divide the read command in two segments as
+		 * recommended by the ESP32 Technical Reference Manual.
+		 */
+		if (msg.len - to_read > 1) {
+			*cmd++ = (struct i2c_esp32_cmd) {
+				.opcode = I2C_ESP32_OP_END
+			};
+		} else {
+			/* Read the last byte and explicitly ask for an
+			 * acknowledgement.
+			 */
+			*cmd++ = (struct i2c_esp32_cmd) {
+				.opcode = I2C_ESP32_OP_READ,
+				.num_bytes = 1,
+				.ack_val = true,
+			};
+
+			/* Account for the `msg.len - 1` when clamping
+			 * transmission length to FIFO buffer size.
+			 */
+			to_read++;
+
+			if (msg.flags & I2C_MSG_STOP) {
+				wait_cmd = cmd;
+				*cmd++ = (struct i2c_esp32_cmd) {
+					.opcode = I2C_ESP32_OP_STOP
+				};
+			}
+		}
+
+		r = i2c_esp32_transmit_and_wait(dev, wait_cmd);
+		if (r < 0) {
+			return r;
+		}
+
+		for (i = 0; i < to_read; i++) {
+			*msg.buf++ = *config->regs.fifo & I2C_FIFO_RDATA;
+		}
+		msg.len -= to_read;
+
+		i2c_esp32_reset_fifo(config);
```

Comment:
Similar to the earlier point: it's generally better to put "hardware reset/cleanup" stuff at the **start** of a command instead of the end.  The software layer usually wants its output as soon as possible and is willing to pay some setup latency.